compiler/rustc_builtin_macros/src/deriving/generic/mod.rs

```diff
@@ -1543,31 +1543,57 @@ impl<'a> TraitDef<'a> {
                             }),
                         ),
                     );
-                    // In general, fields in packed structs are copied via a
-                    // block, e.g. `&{self.0}`. The one exception is `[u8]`
-                    // fields, which cannot be copied and also never cause
-                    // unaligned references. This exception is allowed to
-                    // handle the `FlexZeroSlice` type in the `zerovec` crate
-                    // within `icu4x-0.9.0`.
-                    //
-                    // Once use of `icu4x-0.9.0` has dropped sufficiently, this
-                    // exception should be removed.
-                    let is_u8_slice = if let TyKind::Slice(ty) = &struct_field.ty.kind &&
-                        let TyKind::Path(None, rustc_ast::Path { segments, .. }) = &ty.kind &&
-                        let [seg] = segments.as_slice() &&
-                        seg.ident.name == sym::u8 && seg.args.is_none()
-                    {
-                        true
-                    } else {
-                        false
-                    };
                     if is_packed {
-                        if is_u8_slice {
+                        // In general, fields in packed structs are copied via a
+                        // block, e.g. `&{self.0}`. The two exceptions are `[u8]`
+                        // and `str` fields, which cannot be copied and also never
+                        // cause unaligned references. These exceptions are allowed
+                        // to handle the `FlexZeroSlice` type in the `zerovec`
+                        // crate within `icu4x-0.9.0`.
+                        //
+                        // Once use of `icu4x-0.9.0` has dropped sufficiently, this
+                        // exception should be removed.
+                        //
+                        // Update: unfortunately, after this exception was
+                        // implemented we discovered that `icu4x-0.9.0` still
+                        // doesn't compile even with this exception, due to
+                        // other cases that are harder to work around.
+                        // Therefore these `[u8]` and `str` exceptions should
+                        // be removable relatively quickly, because there are
+                        // no known crates for which they make the difference
+                        // between "the crate compiles" and "the crate does not
+                        // compile".
```

Comment:
This can't be entirely true, can it? The compiler depends on icu4x and that was still building even before https://github.com/unicode-org/icu4x/issues/3060 got fixed. The `[u8]` hack and https://github.com/unicode-org/icu4x/pull/2834 were sufficient for that. Looks like only some configurations of icu4x are affected?

Cc @Manishearth

Reply:
Okay, so there *are* cases where these hacks actually help the build to succeed.

@nnethercote can you adjust the comment? Then r=me.

Reply:
Thanks for the clarification. That makes me feel better about the whole lint :)

Reply:
Yeah, some cases, though in some cases the last field is a ZeroSlice or something, so I don't think a full ICU4X build is fixed by this. However like I said we have a dep update that fixes this so it's not a huge deal.